tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+                         ![msg.to].msgsToLedger = Append(@, msg)
+         ]
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = Tail(subnetMsgs)
+      /\ UNCHANGED<<ledger>>
+      
+LedgerPoll ==
+  \E s \in SUBNETS: Len(subnets[s].msgsToLedger) > 0
+    /\ LET msg == Head(subnets[s].msgsToLedger) IN
+      \* Send message to ledger.
+      /\ ledger' = [ledger EXCEPT !["msgs"] = Append(@,
+          [type |-> "transfer", from |-> msg.from, to |-> msg.to, amount |-> msg.amount]
+        )]
+      \* Remove message from the queue.
+      /\ subnets' = [
+            subnets EXCEPT ![s].msgsToLedger = Tail(@)
+         ]
+      /\ UNCHANGED<<subnetMsgs>>
+
+SubnetReceiveApprove == 
+  /\ Len(subnetMsgs) > 0
+  /\ LET msg == Head(subnetMsgs) IN
+      msg.type = "approve"
+      /\ subnets' = [
+        subnets EXCEPT ![msg.to].unfinalized[msg.from] = @ - msg.amount,
+                       ![msg.to].balance = @ + msg.amount
+      ]
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = Tail(subnetMsgs)
+      /\ UNCHANGED<<ledger>>
+
+LedgerReceiveMessage ==
+  \* If a message is there and the balance checks out, then update it and
+  \* send approval to the subnet. If they don't check out, add sender to
+  \* the dishonestSubnets set and don't update balances.
+  LET
+    msg == Head(ledger.msgs)
+  IN
+   /\ Len(ledger.msgs) > 0
+   
+   /\ IF msg.amount > ledger.balances[msg.from] THEN
+        
+        
+        /\ ledger' = [ledger EXCEPT
+               \* Remove msg from queue.
+               !["msgs"] = Tail(@),
+               
+               \* Subnet spent more than it has. Mark it as a bad subnet.
+               !["dishonestSubnets"] = @ \union {msg.from}
+           ]
+        /\ UNCHANGED<<subnets, subnetMsgs>>
+      ELSE
+        /\ ledger' = [
+            ledger EXCEPT
+              \* Transaction is valid. Update ledger balances.
+              !["balances"][msg.from] = @ - msg.amount,
+              !["balances"][msg.to] = @ + msg.amount,
+              
+              \* Remove msg from queue.
+              !["msgs"] = Tail(@)
+            ]
+        /\ subnetMsgs'= Append(subnetMsgs,
+            [type |-> "approve", from |-> msg.from, to |-> msg.to, amount |-> msg.amount]
+           )
+        /\ UNCHANGED<<subnets>>
+
+Next ==
+  \/ SubnetSendTransfer
+  \/ SubnetReceiveTransfer
+  \/ SubnetReceiveApprove
+  \/ SubnetDishonestSendTransfer
+  \/ LedgerReceiveMessage
+  \/ LedgerPoll
+
+BalancesNonNegative ==
+  (*************************************************************************)
+  (* Invariant to ensure that all subnet balances are non-negative.        *)
+  (*************************************************************************)
+  \A s \in SUBNETS: subnets[s].balance >= 0
+
+
+LedgerIdentifiesAllDishonestSubnets ==
+  {s \in SUBNETS: ~subnets[s].honest} = ledger.dishonestSubnets
+
+LedgerIdentifiesAllDishonestSubnetsEventually ==
```

Comment:
Nit: this property allows the ledger to temporarily falsely suspect a subnet, which may not be desirable (it just says that the ledger _eventually_ doesn't "overshoot" with it's assessment of who's dishonest). If you still want to keep this property, you may want to complement it with an invariant that the set of suspects is a subset of the actually corrupt set.

Reply:
Property removed.